File: flutter_client/lib/data/models/usuario.dart
import 'abstract_model.dart';
import 'palavra.dart';
import 'turma.dart';

class Usuario extends AbstractModel {
  String? nome;
  String? username;
  String? codigoIdentificador;
  String? password;
  String? email;
  String? avatar;
  String? role;
  String dificuldade;
  bool mustChangePassword;
  List<Palavra>? palavras;
  List<Turma>? turmas;

  Usuario({
    super.id,
    super.createdAt,
    this.nome,
    this.username,
    this.codigoIdentificador,
    this.password,
    this.email,
    this.avatar,
    this.role,
    this.dificuldade = 'FACIL',
    this.mustChangePassword = false,
    this.palavras,
    this.turmas,
  });

  bool get isAdmin => role?.toUpperCase() == 'ADMIN';

  factory Usuario.fromJson(Map<String, dynamic> json) {
    return Usuario(
      id: json['id'] as int?,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      nome: json['nome'] as String?,
      username: json['username'] as String?,
      codigoIdentificador: json['codigoIdentificador'] as String?,
      password: json['password'] as String?,
      email: json['email'] as String?,
      avatar: json['avatar'] as String? ?? 'assets/avatar/avatar_1.jpg',
      role: json['role'] as String? ?? 'USER',
      dificuldade: json['dificuldade'] as String? ?? 'FACIL',
      mustChangePassword: json['mustChangePassword'] as bool? ?? false,
      palavras: json['palavras'] != null
          ? (json['palavras'] as List).map((e) => Palavra.fromJson(e)).toList()
          : null,
      turmas: json['turmas'] != null
          ? (json['turmas'] as List).map((e) => Turma.fromJson(e)).toList()
          : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'createdAt': createdAt?.toIso8601String(),
      'nome': nome,
      'username': username,
      'codigoIdentificador': codigoIdentificador,
      'password': password,
      'email': email,
      'avatar': avatar ?? 'assets/avatar/avatar_1.jpg',
      'role': role ?? 'USER',
      'dificuldade': dificuldade,
      'mustChangePassword': mustChangePassword,
      'palavras': palavras?.map((e) => e.toJson()).toList(),
      'turmas': turmas?.map((e) => e.toJson()).toList(),
    };
  }
}

File: server_java/src/main/java/br/com/interalibras/entity/Usuario.java
package br.com.interalibras.entity;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "usuarios")
public class Usuario {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String nome;

    @Column(unique = true, nullable = true)
    private String email;

    @Column(unique = true, nullable = false)
    private String username;

    @Column(unique = true, nullable = true)
    private String codigoIdentificador;

    @Column(nullable = false)
    private String password;

    private String avatar = "assets/avatar/avatar_1.jpg";

    private String role = "USER";

    private Boolean mustChangePassword = false;

    private LocalDateTime createdAt = LocalDateTime.now();

    public Usuario() {}

    public Usuario(Long id, String nome, String email, String username, String password) {
        this.id = id;
        this.nome = nome;
        this.email = email;
        this.username = username;
        this.password = password;
        this.avatar = "assets/avatar/avatar_1.jpg";
        this.role = "USER";
        this.mustChangePassword = false;
    }

    public Usuario(Long id, String nome, String email, String username, String codigoIdentificador, String password, String avatar, String role) {
        this.id = id;
        this.nome = nome;
        this.email = email;
        this.username = username;
        this.codigoIdentificador = codigoIdentificador;
        this.password = password;
        this.avatar = avatar != null ? avatar : "assets/avatar/avatar_1.jpg";
        this.role = role != null ? role : "USER";
        this.mustChangePassword = false;
    }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getNome() { return nome; }
    public void setNome(String nome) { this.nome = nome; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getUsername() { return username; }
    public void setUsername(String username) { this.username = username; }

    public String getCodigoIdentificador() { return codigoIdentificador; }
    public void setCodigoIdentificador(String codigoIdentificador) { this.codigoIdentificador = codigoIdentificador; }

    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }

    public String getAvatar() { return avatar; }
    public void setAvatar(String avatar) { this.avatar = avatar; }

    public String getRole() { return role; }
    public void setRole(String role) { this.role = role; }

    public Boolean getMustChangePassword() { return mustChangePassword != null ? mustChangePassword : false; }
    public void setMustChangePassword(Boolean mustChangePassword) { this.mustChangePassword = mustChangePassword; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
}
